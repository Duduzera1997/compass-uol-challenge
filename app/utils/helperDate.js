import { format, formatDistanceToNow as formatDistance } from 'date-fns';
import ptBR from 'date-fns/locale/pt-BR';

export function formatDateLiteral(date) {
  return format(new Date(date), "dd 'de' MMMM 'de' yyyy", {
    locale: ptBR,
  });
}

export function formatDistanceToNow(date) {
  return `${formatDistance(date, { locale: ptBR })} atr√°s`;
}

export function formatDateWithDistance(date) {
  return `${formatDateLiteral(date)} | ${formatDistanceToNow(date)}`;
}
