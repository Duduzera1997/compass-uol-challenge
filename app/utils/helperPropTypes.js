import PropTypes from 'prop-types';

export const ArrayOfArticlePropTypes = PropTypes.shape({
  last_updated: PropTypes.string,
  num_results: PropTypes.number,
  results: PropTypes.arrayOf(PropTypes.shape(ArticlePropTypes)).isRequired,
});

export const ArticlePropTypes = PropTypes.shape({
  section: PropTypes.string,
  subsection: PropTypes.string,
  title: PropTypes.string,
  abstract: PropTypes.string,
  url: PropTypes.string,
  uri: PropTypes.string,
  byline: PropTypes.string,
  item_type: PropTypes.string,
  updated_date: PropTypes.string,
  created_date: PropTypes.string,
  published_date: PropTypes.string,
  material_type_facet: PropTypes.string,
  kicker: PropTypes.string,
  des_facet: PropTypes.arrayOf(PropTypes.string),
  org_facet: PropTypes.arrayOf(PropTypes.string),
  per_facet: PropTypes.arrayOf(PropTypes.string),
  geo_facet: PropTypes.arrayOf(PropTypes.string),
  multimedia: PropTypes.arrayOf(PropTypes.shape(MultimidiaPropTypes)),
  short_url: PropTypes.string,
});

export const MultimidiaPropTypes = PropTypes.shape({
  url: PropTypes.string,
  format: PropTypes.string,
  height: PropTypes.integer,
  width: PropTypes.integer,
  type: PropTypes.string,
  subtype: PropTypes.string,
  caption: PropTypes.string,
  copyright: PropTypes.string,
});
