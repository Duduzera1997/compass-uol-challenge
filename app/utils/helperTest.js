import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider } from 'styled-components/native';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import theme from '~/styles/theme';

export const renderWithTheme = (component) =>
  render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);

export const renderWithStoreAndThemeProviders = (component, { techMock }) => {
  const mockStore = configureStore();
  const state = {
    techReducer: {
      ...techMock,
    },
  };
  const store = mockStore(state);
  return renderWithTheme(<Provider store={store}>{component}</Provider>);
};
