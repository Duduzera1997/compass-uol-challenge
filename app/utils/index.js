export * from './constants';
export * from './helperDate';
export * from './helperTest';
export * from './helperPropTypes';
