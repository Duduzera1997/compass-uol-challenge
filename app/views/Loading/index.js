import React from 'react';
import { Loading } from '~/components';

function LoadingScreen() {
  return <Loading />;
}
export default LoadingScreen;
