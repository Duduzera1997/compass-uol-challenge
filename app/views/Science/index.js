import React from 'react';
import { View, Text } from 'react-native';

function ScienceScreen() {
  return (
    <View>
      <Text>Science</Text>
    </View>
  );
}

export default ScienceScreen;
