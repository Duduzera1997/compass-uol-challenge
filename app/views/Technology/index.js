import React, { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import _ from 'lodash';

import * as SC from './styles';
import { techActions } from '~/redux';
import {
  LinearBackground,
  Header,
  ListArticles,
  ViewArticleModal,
  SearchInput,
} from '~/components';

function TechnologyScreen() {
  const { data, loading, error } = useSelector((state) => state.techReducer);
  const [filterText, setFilterText] = useState('');

  const [modalInfo, setModalInfo] = useState({ visible: false, item: null });
  const dispatch = useDispatch();

  useEffect(() => {
    fetchArticlesTechnology();
  }, [fetchArticlesTechnology]);

  const fetchArticlesTechnology = useCallback(
    () => dispatch(techActions.fetchTechnologiesAction()),
    [dispatch],
  );

  function getFilteredList() {
    if (_.isEmpty(filterText)) return data;
    const filteredResults = data.results.filter(
      (item) =>
        item.title.toLowerCase().includes(filterText.toLowerCase()) ||
        item.abstract.toLowerCase().includes(filterText.toLowerCase()),
    );
    return {
      ...data,
      results: filteredResults,
      num_results: filteredResults.length,
    };
  }

  function onCardPressed(item) {
    setModalInfo({ ...modalInfo, visible: true, item });
  }

  function onRequestCloseModal() {
    setModalInfo({ ...modalInfo, visible: false, item: null });
  }

  const filteredList = getFilteredList();

  return (
    <LinearBackground>
      <SC.ContainerTech>
        {modalInfo.visible && (
          <ViewArticleModal
            visible={modalInfo.visible}
            item={modalInfo.item}
            onRequestClose={onRequestCloseModal}
          />
        )}
        <SC.ContainerByFlex flex={1}>
          <SC.ContainerByFlex flex={1}>
            <Header headerText="Technology" />
          </SC.ContainerByFlex>
          {!!data?.results?.length && (
            <SC.ContainerByFlex flex={1}>
              <SearchInput onChangeText={_.debounce(setFilterText, 500)} />
            </SC.ContainerByFlex>
          )}
          <SC.ContainerByFlex flex={6}>
            <ListArticles
              data={filteredList}
              loading={
                loading || (!loading && !!filterText && !!filteredList.length)
              }
              error={error}
              onRefresh={fetchArticlesTechnology}
              onCardPress={onCardPressed}
            />
          </SC.ContainerByFlex>
        </SC.ContainerByFlex>
      </SC.ContainerTech>
    </LinearBackground>
  );
}

export default TechnologyScreen;
