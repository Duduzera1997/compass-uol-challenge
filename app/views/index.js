import LoadingScreen from './Loading';
import TechnologyScreen from './Technology';
import ScienceScreen from './Science';

export { LoadingScreen, TechnologyScreen, ScienceScreen };
