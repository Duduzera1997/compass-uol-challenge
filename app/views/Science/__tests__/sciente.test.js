import React from 'react';

import { ScienceScreen } from '~/views';
import { renderWithStoreAndThemeProviders } from '~/utils';

const mock = {
  section: 'Science',
  last_updated: '2021-03-19T17:10:56-04:00',
  num_results: 1,
  results: [
    {
      title: 'How to Vaccinate Homebound Seniors? Take the Shots to Them.',
      abstract:
        'Millions of older Americans, chronically ill and trapped at home, pose a big challenge to inoculation efforts. Doctors are on it.',
      url:
        'https://www.nytimes.com/2021/03/22/health/virus-seniors-vaccination.html',
      uri: 'nyt://article/e31cffee-6e8f-54eb-95dc-3ab6337fdb18',
      byline: 'By Roni Caryn Rabin and Chang W. Lee',
      item_type: 'Article',
      published_date: '2021-03-22T17:08:27-04:00',
      multimedia: [
        {
          url:
            'https://static01.nyt.com/images/2021/03/22/science/00VIRUS-HOMEBOUND-promo/00VIRUS-HOMEBOUND-promo-articleInline.jpg',
        },
      ],
    },
  ],
};

describe('Science Screen', () => {
  it('should be contains a Science text Label', () => {
    const state = {
      data: mock,
      loading: false,
      error: false,
    };

    const { getByText, toJSON } = renderWithStoreAndThemeProviders(
      <ScienceScreen />,
      {
        scienceMock: state,
      },
    );
    const component = getByText(/Science/);
    expect(component).toBeDefined();
    expect(toJSON()).toMatchSnapshot();
  });
});
