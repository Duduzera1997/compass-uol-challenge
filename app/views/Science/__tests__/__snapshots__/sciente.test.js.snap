// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Science Screen should be contains a Science text Label 1`] = `
<BVLinearGradient
  colors={
    Array [
      4282066552,
      4294924895,
    ]
  }
  endPoint={
    Object {
      "x": 0,
      "y": 1.05,
    }
  }
  locations={null}
  startPoint={
    Object {
      "x": 0,
      "y": 0,
    }
  }
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingTop": 10,
        },
      ]
    }
  >
    <View
      flex={1}
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
    >
      <View
        flex={1}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-start",
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "PTSansNarrowBold",
                  "fontSize": 27,
                  "letterSpacing": 2,
                  "paddingLeft": 15,
                  "paddingTop": 15,
                },
              ]
            }
          >
            Science
             News
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "PTSansNarrowRegular",
                  "fontSize": 14,
                  "paddingLeft": 16,
                  "paddingTop": 2,
                },
              ]
            }
          >
            Hoje é, 
            23 de março de 2021
            .
          </Text>
        </View>
      </View>
      <View
        flex={1}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          placeholder="Busque uma notícia..."
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "backgroundColor": "#FFFFFF",
                "borderColor": "#FFFFFF",
                "borderRadius": 9,
                "borderWidth": 1,
                "color": "#333333",
                "fontFamily": "PTSansNarrowRegular",
                "fontSize": 16,
                "marginBottom": 0,
                "marginLeft": 16,
                "marginRight": 0,
                "marginTop": 10,
                "paddingBottom": 10,
                "paddingLeft": 10,
                "paddingRight": 10,
                "paddingTop": 10,
                "width": "92%",
              },
            ]
          }
          underlineColorAndroid="transparent"
        />
      </View>
      <View
        flex={6}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 6,
              "flexShrink": 1,
            },
          ]
        }
      >
        <RCTScrollView
          ListEmptyComponent={
            <EmptyList
              isLoading={false}
            />
          }
          contentContainerStyle={
            Object {
              "paddingBottom": 20,
              "paddingLeft": 20,
              "paddingRight": 20,
            }
          }
          data={
            Array [
              Object {
                "abstract": "Millions of older Americans, chronically ill and trapped at home, pose a big challenge to inoculation efforts. Doctors are on it.",
                "byline": "By Roni Caryn Rabin and Chang W. Lee",
                "item_type": "Article",
                "multimedia": Array [
                  Object {
                    "url": "https://static01.nyt.com/images/2021/03/22/science/00VIRUS-HOMEBOUND-promo/00VIRUS-HOMEBOUND-promo-articleInline.jpg",
                  },
                ],
                "published_date": "2021-03-22T17:08:27-04:00",
                "title": "How to Vaccinate Homebound Seniors? Take the Shots to Them.",
                "uri": "nyt://article/e31cffee-6e8f-54eb-95dc-3ab6337fdb18",
                "url": "https://www.nytimes.com/2021/03/22/health/virus-seniors-vaccination.html",
              },
            ]
          }
          disableVirtualization={false}
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          keyboardShouldPersistTaps="never"
          maxToRenderPerBatch={10}
          onContentSizeChange={[Function]}
          onEndReached={[Function]}
          onEndReachedThreshold={0.1}
          onLayout={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          refreshControl={
            <RefreshControlMock
              enabled={true}
              onRefresh={[Function]}
              refreshing={false}
            />
          }
          renderItem={[Function]}
          scrollEventThrottle={50}
          showsVerticalScrollIndicator={false}
          stickyHeaderIndices={Array []}
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
          updateCellsBatchingPeriod={50}
          windowSize={21}
        >
          <RCTRefreshControl />
          <View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "height": 120,
                      "paddingBottom": 10,
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 9,
                      "flexDirection": "row",
                      "height": "100%",
                      "opacity": 1,
                    }
                  }
                >
                  <View
                    flex={1}
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexBasis": 0,
                          "flexGrow": 1,
                          "flexShrink": 1,
                          "justifyContent": "flex-start",
                          "paddingBottom": 2,
                          "paddingLeft": 7,
                          "paddingRight": 5,
                          "paddingTop": 5,
                        },
                      ]
                    }
                  >
                    <Image
                      source={
                        Object {
                          "uri": "https://static01.nyt.com/images/2021/03/22/science/00VIRUS-HOMEBOUND-promo/00VIRUS-HOMEBOUND-promo-articleInline.jpg",
                        }
                      }
                      style={
                        Array [
                          Object {
                            "borderRadius": 10,
                            "height": "95%",
                            "width": "100%",
                          },
                        ]
                      }
                    />
                  </View>
                  <View
                    flex={2.5}
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexBasis": 0,
                          "flexGrow": 2.5,
                          "flexShrink": 1,
                          "justifyContent": "flex-start",
                          "paddingBottom": 2,
                          "paddingLeft": 7,
                          "paddingRight": 5,
                          "paddingTop": 5,
                        },
                      ]
                    }
                  >
                    <Text
                      fontBold={true}
                      numberOfLines={2}
                      size={15}
                      style={
                        Array [
                          Object {
                            "color": "#333333",
                            "fontFamily": "PTSansNarrowBold",
                            "fontSize": 15,
                            "paddingBottom": 2,
                            "paddingLeft": 4,
                            "paddingRight": 2,
                            "paddingTop": 4,
                            "width": "99%",
                          },
                        ]
                      }
                    >
                      How to Vaccinate Homebound Seniors? Take the Shots to Them.
                    </Text>
                    <Text
                      fontBold={true}
                      size={11.5}
                      style={
                        Array [
                          Object {
                            "color": "#767676",
                            "fontFamily": "PTSansNarrowBold",
                            "fontSize": 11.5,
                            "paddingBottom": 2,
                            "paddingLeft": 4,
                            "paddingRight": 2,
                            "paddingTop": 0,
                            "width": "99%",
                          },
                        ]
                      }
                    >
                      22 de março de 2021
                    </Text>
                    <Text
                      numberOfLines={2}
                      size={13}
                      style={
                        Array [
                          Object {
                            "color": "#333333",
                            "fontFamily": "PTSansNarrowRegular",
                            "fontSize": 13,
                            "paddingBottom": 2,
                            "paddingLeft": 4,
                            "paddingRight": 2,
                            "paddingTop": 4,
                            "width": "99%",
                          },
                        ]
                      }
                    >
                      Millions of older Americans, chronically ill and trapped at home, pose a big challenge to inoculation efforts. Doctors are on it.
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </View>
</BVLinearGradient>
`;
